hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HStore.java

```diff
@@ -1088,9 +1086,19 @@ public HStoreFile tryCommitRecoveredHFile(Path path) throws IOException {
    * @return store file created.
    */
   private HStoreFile commitFile(Path path, long logCacheFlushId, MonitoredTask status)
-      throws IOException {
-    // Write-out finished successfully, move into the right spot
-    Path dstPath = getRegionFileSystem().commitStoreFile(getColumnFamilyName(), path);
+    throws IOException {
+    HRegionFileSystem hfs = getRegionFileSystem();
+    Path storeDir = hfs.getStoreDir(getColumnFamilyName());
+    Path dstPath;
+    // As we want to support writing to data directory directly, here we need to check whether the
+    // store file is already in the right place
+    if (path.getParent() != null && path.getParent().equals(storeDir)) {
+      // already in the right place, skip renmaing
+      dstPath = path;
+    } else {
+      // Write-out finished successfully, move into the right spot
+      dstPath = hfs.commitStoreFile(getColumnFamilyName(), path);
+    }
```

Comment:
nit: As mentioned in the design doc, this could be delegated to SFT implementations (similarly to how createWriter was moved from Compactor/StoreFlusher to SFT).